Add `optimizeDirectReadFromTextIndex` information to the debug level logs for testing

Ergus opened the pull request:
This PR add `optimizeDirectReadFromTextIndex` information to the debug level logs.
That information is used to test the direct read optimization transformations.


### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Ergus committed:
Add log message and initial test

rschu1ze commented on src/Processors/QueryPlan/Optimizations/optimizeDirectReadFromTextIndex.cpp:
```diff
@@ -36,6 +36,39 @@ String getNameWithoutAliases(const ActionsDAG::Node * node)
+
+    String result = "Added: [";
+
+    // This will list the index and the new associated columns
```
Minor: `//` --> `///`

rschu1ze commented on src/Processors/QueryPlan/Optimizations/optimizeDirectReadFromTextIndex.cpp:
```diff
@@ -36,6 +36,39 @@ String getNameWithoutAliases(const ActionsDAG::Node * node)
     return node->result_name;
 }
 
+String optimizationInfoToString(const IndexReadColumns &added_columns, const Names &removed_columns)
```
Minor: ` & ` (space left and right of &)

rschu1ze commented on src/Processors/QueryPlan/Optimizations/optimizeDirectReadFromTextIndex.cpp:
```diff
@@ -36,6 +36,39 @@ String getNameWithoutAliases(const ActionsDAG::Node * node)
+
+    // This will list the index and the new associated columns
+    size_t idx = 0;
+    for (const auto &[_, name_and_type] : added_columns)
```
Minor: Spaces around &, please.

rschu1ze replied:
This must be plural and less ambiguous. `name_and_type` --> `columns_names_and_types`

rschu1ze commented on src/Processors/QueryPlan/Optimizations/optimizeDirectReadFromTextIndex.cpp:
```diff
@@ -36,6 +36,39 @@ String getNameWithoutAliases(const ActionsDAG::Node * node)
+    String result = "Added: [";
+
+    // This will list the index and the new associated columns
+    size_t idx = 0;
```
`bool first_iteration = true`

Ergus replied:
Solved via e77a72d2c6edfd43227a8c

rschu1ze commented on src/Processors/QueryPlan/Optimizations/optimizeDirectReadFromTextIndex.cpp:
```diff
@@ -36,6 +36,39 @@ String getNameWithoutAliases(const ActionsDAG::Node * node)
+    size_t idx = 0;
+    for (const auto &[_, name_and_type] : added_columns)
+    {
+        for (const String & column_name : name_and_type.getNames())
```
`getNames()` returns a vector. I don't quite understand why we need variable `idx` to keep track of indexes when we could iterate the elements like in l. 61.

rschu1ze commented on tests/queries/0_stateless/02346_text_index_direct_read.sql:
```diff
@@ -0,0 +1,108 @@
+
+SET log_queries = 1;
+
+-- Affects the number of read rows.
```
The comment in l. 6 applies to l. 8

rschu1ze commented on tests/queries/0_stateless/02346_text_index_direct_read.sql:
```diff
@@ -0,0 +1,108 @@
+SET allow_experimental_full_text_index = 1;
+SET use_skip_indexes_on_data_read = 1;
+SET query_plan_direct_read_from_text_index = 1;
+SET max_rows_to_read = 0; -- system.text_log can be really big
```
Hmm, so `max_rows_to_read` will make all SELECTs in this no test at all, right?

rschu1ze commented on tests/queries/0_stateless/02346_text_index_direct_read.sql:
```diff
@@ -0,0 +1,108 @@
+
+SYSTEM FLUSH LOGS text_log;
+
+SELECT message FROM system.text_log WHERE
```
`FLUSH LOGS` does not _clear_ the log. We may see a leftover entry from the SELECT in l. 43 in the text_log.

rschu1ze commented on tests/queries/0_stateless/02346_text_index_direct_read.sql:
```diff
@@ -0,0 +1,108 @@
+       startsWith(message, 'Added:') ORDER BY event_time_microseconds DESC LIMIT 1;
+
+
+DROP TABLE IF EXISTS tab;
```
While at it, could you please extend the "direct read" section in the [docs](https://clickhouse.com/docs/engines/table-engines/mergetree-family/invertedindexes#direct-read)? It needs
- a high-level description of direct read,
- a list of supported functions,
- a note how it plays together with AND/OR/NOT and/or other (potentially non-text search) functions.

Ergus replied:
Solved via:  dbc38e42988e29cae9ffb79d1e

Please specify if you prefer to add more examples or more details in the text.

Ergus committed:
Add NOT test

Ergus committed:
CR @rschu1ze

rschu1ze commented on src/Processors/QueryPlan/Optimizations/optimizeDirectReadFromTextIndex.cpp:
```diff
@@ -36,6 +36,39 @@ String getNameWithoutAliases(const ActionsDAG::Node * node)
+    {
+        for (const String & column_name : columns_names_and_types.getNames())
+        {
+            if (!std::exchange(first_iteration, false))
```
Looks un-pretty, let's revert back to the previous state.

Ergus committed:
Improve documentation

Ergus committed:
Reuse a counter to track first entry.

Ergus committed:
Fix markdown linter error

Ergus committed:
Merge remote-tracking branch 'principal/master' into test_direct_read_plan

Ergus committed:
Fix test performance issue

The test was taking ~20 seconds because the log flush+ text_log parse is very expensive.
This change makes testing time ~3 seconds

rschu1ze committed:
Update docs

rschu1ze commented on tests/queries/0_stateless/02346_text_index_direct_read.sql:
```diff
@@ -0,0 +1,50 @@
+-- Tags: no-fasttest
+-- no-fasttest: It can be slow
```
Are you sure? This test looks like it runs in a couple of seconds. The only slow operation may the logs flushing.

If other similar tests which check `system.text_log` also don't run in FastTests, then we can exclude this test too, otherwise it will be nicer to remove l. 1-2.

Ergus replied:
Fixed. Previously it needed ~20 seconds to run, but after [6eddda7](https://github.com/ClickHouse/ClickHouse/pull/87955/commits/6eddda7a9c0e19d35c85e0ebcc0f8b2c50edb17f) it only takes ~2 seconds.

Ergus committed:
Add no-parallel to the test to avoid text_log concurrency issues

Ergus committed:
Merge remote-tracking branch 'principal/master' into test_direct_read_plan

Ergus committed:
Remove redundant IF EXIST form test

Ergus committed:
Merge remote-tracking branch 'principal/master' into test_direct_read_plan

Ergus committed:
Add  no-parallel-replicas to the test

Ergus committed:
Improve filter_column replacement.

After an optimization and removing unused nodes the output nodes may change order.
In that case the first output is not guaranteed to be the filter_column.
Similarly if some previous optimization reordered the output nodes after constructing the filter_step.

rschu1ze commented:
[Stateless tests (amd_msan, sequential, 1/2)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=87955&sha=latest&name_0=PR&name_1=Stateless+tests+%28amd_msan%2C+sequential%2C+1%2F2%29): https://github.com/ClickHouse/ClickHouse/issues/88107

Ergus committed:
Merge remote-tracking branch 'principal/master' into test_direct_read_plan

rschu1ze commented on tests/queries/0_stateless/02346_text_index_direct_read.sql:
```diff
@@ -0,0 +1,51 @@
+
+DROP TABLE IF EXISTS tab;
+
+CREATE TABLE tab(k UInt64, text String, INDEX idx(text) TYPE text(tokenizer = 'default') GRANULARITY 1)
```
We need to use the new tokenizer names from Elmis PR here.

Ergus committed:
Use new syntax for the tokenizer in text index.

Update test 02346_text_index_direct_read.sql
